File: tvr-decoder/src/main/java/com/willcurrie/controllers/HexDumpFactory.java
package com.willcurrie.controllers;

import com.willcurrie.hex.ByteElement;
import com.willcurrie.hex.HexDumpElement;
import com.willcurrie.hex.WhitespaceElement;

import java.util.ArrayList;
import java.util.List;

public class HexDumpFactory {
    public List<HexDumpElement> splitIntoByteLengthStrings(String hexString, int startIndexInBytes) {
        List<HexDumpElement> elements = new ArrayList<HexDumpElement>();
        int byteOffset = startIndexInBytes;
        for (int i = 0; i < hexString.length(); ) {
            if (hexString.charAt(i) == ' ') {
                elements.add(new WhitespaceElement("<br>"));
                i++;
            } else {
                elements.add(new ByteElement(hexString.substring(i, i + 2), byteOffset++));
                i += 2;
            }
        }
        return elements;
    }
}

File: src/main/java/io/github/binaryfoo/HexDumpFactory.kt
package io.github.binaryfoo

import io.github.binaryfoo.hex.ByteElement
import io.github.binaryfoo.hex.HexDumpElement
import io.github.binaryfoo.hex.WhitespaceElement

import java.util.ArrayList

public class HexDumpFactory {
    public fun splitIntoByteLengthStrings(hexString: String, startIndexInBytes: Int): List<HexDumpElement> {
        val elements = ArrayList<HexDumpElement>()
        var byteOffset = startIndexInBytes
        for (i in 0..hexString.length-2 step 2) {
            elements.add(ByteElement(hexString.substring(i, i + 2), byteOffset++))
        }
        return elements
    }
}